src/basicfuns.jl

```diff
@@ -152,9 +152,47 @@ Return `log(1+exp(x))` evaluated carefully for largish `x`.
 
 This is also called the ["softplus"](https://en.wikipedia.org/wiki/Rectifier_(neural_networks))
 transformation, being a smooth approximation to `max(0,x)`. Its inverse is [`logexpm1`](@ref).
+
+See:
+ * Martin Maechler (2012) [“Accurately Computing log(1 − exp(− |a|))”](http://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf)
 """
-log1pexp(x::Real) = x < 18.0 ? log1p(exp(x)) : x < 33.3 ? x + exp(-x) : oftype(exp(-x), x)
-log1pexp(x::Float32) = x < 9.0f0 ? log1p(exp(x)) : x < 16.0f0 ? x + exp(-x) : oftype(exp(-x), x)
+function log1pexp(_x::Real)
+    x = float(_x)
+    x0, x1, x2 = _log1pexp_thresholds(x)
+    if x < x0
+        return exp(x)
+    elseif x < x1
+        return log1p(exp(x))
+    elseif x < x2
+        return x + exp(-x)
+    else
+        return x
+    end
+end
+
+#=
+Returns a thresholds x0, x1, x2 such that:
+
+    * log1pexp(x) ≈ exp(x) for x ≤ x0
+    * log1pexp(x) ≈ log1p(exp(x)) for x0 < x ≤ x1
+    * log1pexp(x) ≈ x + exp(-x) for x1 < x ≤ x2
+    * log1pexp(x) ≈ x for x > x2
+
+where the tolerances of the approximations ≈ are on the order of eps(T).
+=#
+@inline @generated function _log1pexp_thresholds(::T) where {T<:Real}
+    ϵ = big(eps(T))
+    x0 = log(ϵ / 2)
+    x1 = -log(ϵ) / 2
+    x2 = -x0 - log(-x0) * (1 + 1 / x0) # ≈ root of e^-x == x * ϵ/2 via asymptotics of Lambert's W function
+    return (T(x0), T(x1), T(x2))
+end
```

Comment:
One could also use a `if @generated` block to let the compiler decide if it uses the `@generated` version or not (https://docs.julialang.org/en/v1/manual/metaprogramming/#Optionally-generated-functions).

Reply:
> One could also use a `if @generated` block to let the compiler decide if it uses the `@generated` version or not (https://docs.julialang.org/en/v1/manual/metaprogramming/#Optionally-generated-functions).

What's the benefit of this?

Reply:
It can reduce compile time cost:
> Generated functions can achieve high efficiency at run time, but come with a compile time cost: a new function body must be generated for every combination of concrete argument types. Typically, Julia is able to compile "generic" versions of functions that will work for any arguments, but with generated functions this is impossible. This means that programs making heavy use of generated functions might be impossible to statically compile.
>
> To solve this problem, the language provides syntax for writing normal, non-generated alternative implementations of generated functions.